### PR TITLE
Faster integer hashing (fixes #37800 UB)

### LINKED ISSUE #37800: Floating point hash bug
```
using InteractiveUtils
using Printf

function myhash(x)
    # the following is valid for x == Inf
    h = UInt(0)
    num, pow, den = Base.decompose(x)
    den == 0 && return hash(ifelse(num > 0, Inf, -Inf), h)
    return nothing
end

versioninfo()
h = hash(Inf)
@printf("%#x\n", h)
h = myhash(Inf)
@printf("%#x\n", h)
```
With LLVM 9 I get the expected output:
```
Julia Version 1.6.0-DEV.1014
Commit 2a36f83ce9 (2020-09-22 20:52 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
0x807bb202c9cbfc6f
0x807bb202c9cbfc6f
```
With LLVM 10 I get:
```
Julia Version 1.6.0-DEV.1015
Commit 9adcdd42a0 (2020-09-23 03:04 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-10.0.1 (ORCJIT, skylake)
0x807bb202c9cbfc6f
0x15d7d083d04ecb90
```
Edit:
`0x15d7d083d04ecb90` is the correct hash for `NaN`

### PR DESCRIPTION
**Pull request by vtjnash**
In LLVM (inherited from C), fptosi has undefined behavior if the result
does not fit the integer size after rounding down. But by using the same
strategy as generic hashing of Real values, we actually can end up with
a situation that is not much slower for the CPU to deal with and avoids
the UB.

Refs #6624 (36969687fb9bbd936b3adc0baa6cf0a408d2ccb7)
Fixes #37800

Benchmarks:
```
julia> Random.seed!(1); a = rand(Int64, 10^8); b = copy(reinterpret(Float64, a)); c = rand(Float64, length(a)); d = copy(reinterpret(UInt64, a));
julia> @time sum(hash, a); @time sum(hash, b); @time sum(hash, c); @time sum(hash, d);
```
```
master:
  0.174493 seconds (1 allocation: 16 bytes)
  0.316409 seconds (1 allocation: 16 bytes)
  0.317264 seconds (1 allocation: 16 bytes)
  0.145938 seconds (1 allocation: 16 bytes)

PR:
  0.119815 seconds (1 allocation: 16 bytes)
  0.687834 seconds (1 allocation: 16 bytes)
  0.468544 seconds (1 allocation: 16 bytes)
  0.210141 seconds (1 allocation: 16 bytes)
```
Aside: note that integers-as-floating-point aren't affected as much:
```
julia> e = map(Float64, a);
master:  0.313309 seconds (1 allocation: 16 bytes)
PR:   0.338984 seconds (1 allocation: 16 bytes)
```

@StefanKarpinski I picked these factors mostly at random. So if you have better ideas for the specific hashing function itself, I'll update to that. This is showing the general idea and testing performance of it.

### CODE REVIEW

**Comment by jmkuhn** [comment]
With this PR you have
```
julia> hash(0.0) == hash(-0.0)
true
```

**Comment by vtjnash** [comment]
Ah, you're right. Good catch. I completely botched that and forgot to use `isequal` with hashing instead of `==`.

File: base/float.jl [hunk 1]
```diff
@@ -551,12 +551,32 @@ isinf(x::Real) = !isnan(x) & !isfinite(x)
...
+            end
+        elseif $(Tf(typemin(Tu))) <= x < $(Tf(typemax(Tu)))
+            xu = fptoui($Tu, x)
+            if isequal(xu, x)
+                return hash(xu, h)
+            end
+        elseif isnan(x)
+            return hx_NaN ⊻ h # NaN does not have a stable bit pattern
+        end
+        return hash_uint64(5bitcast(UInt64, x)) - 3h
```
Comment (StefanKarpinski) [inline base/float.jl:577]:
```suggestion
        return hash_uint64(5*bitcast(UInt64, x)) - 3h
```

**Comment by vtjnash** [comment]
triage concluded to simplify this (disable the breaking tests), and otherwise to proceed

**Comment by Keno** [comment]
What is holding this up? IIRC Triage ok'ed this, so can we just merge it?

**Pull request merged**